### PR TITLE
SOL-121556: Golang: Implement code changes to support the endpoint provision feature

### PR DESCRIPTION
**Pull request by oodigie**
**Changes:**
- SOL-121556: Golang: Implement code changes to support the endpoint provision feature
- feat[SOL-121556]: fixed failing tests
- SOL-121557: Golang: Implement the endpoint deprovision  feature
- chore: removed some unnecessary comments from some tests
- feat[SOL-121560]: wrote integration tests to provide 100% coverage for the new code changes

### CODE REVIEW

**Comment by cjwmorgan-sol** [comment]
Looks like there are statichceck and got vet failures witht he source please check the git actions.

**Comment by oodigie** [comment]
> Looks like there are statichceck and got vet failures witht he source please check the git actions.

Yeah, I have fixed it. Thanks

**Review by cjwmorgan-sol — CHANGES_REQUESTED** [review]
Mostly looking good.
I have some concerns about the default endpoint properties, as they prevent broker defaults from being used.
Made a few suggestions for adding or enhancing debug logs. Up for discussion if not trivial to implement.
Git actions show all new unit tests pass.
Look to comments to address any changes.

File: internal/impl/constants/default_properties.go
Comment (sdewilde) [reply internal/impl/constants/default_properties.go:76]: +1 to what @cjwmorgan-sol mentioned, we should only be setting the durability be default.  Included a screenshot of Go with no props set by the user (left) vs CCSMP with no props (right)
image

File: pkg/solace/config/endpoint_properties_test.go
Comment (sdewilde) [inline pkg/solace/config/endpoint_properties_test.go:27]: Is there a reason for not testing all of the props we have?

File: internal/impl/provisioner/endpoint_provisioner_impl_test.go
Comment (sdewilde) [inline internal/impl/provisioner/endpoint_provisioner_impl_test.go:45]: Typo in the comments?

File: internal/impl/constants/default_properties.go
Comment (oodigie) [reply internal/impl/constants/default_properties.go:76]: I will update this.

File: pkg/solace/endpoint_provisioner.go
Comment (oodigie) [reply pkg/solace/endpoint_provisioner.go:133]: Okay

File: pkg/solace/config/endpoint_properties_test.go
Comment (oodigie) [reply pkg/solace/config/endpoint_properties_test.go:27]: yes. The remaining props are either boolean or uint values and since Go is a strongly typed language, there is no need to have unit tests for boolean/unit values (I cannot test an invalid boolean/unit argument). The other testing is covered by integration tests that will come in another PR.

Comment (sdewilde) [reply pkg/solace/config/endpoint_properties_test.go:27]: This test isn't testing invalid values though, it is the serialization/deserialization
Even in this test `EndpointPropertyMaxMessageRedelivery` is being set as a float in your test but it should be a uint in the actual type...

File: pkg/solace/endpoint_provisioner.go
Comment (sdewilde) [inline pkg/solace/endpoint_provisioner.go:120]: Comment mentions that this outcome is used in the prov and deprov operations but we are only returning an `error` from deprovision.  The arch agrees with your implementation not this comment but after writing some code with this I'd prefer they both return the same type (or channel of the same type) 
@cjwmorgan-sol thoughts here? Not sure if this has big impacts on other areas of deprovision is we change it form just an error

File: pkg/solace/endpoint_provisioner.go
Comment (oodigie) [reply pkg/solace/endpoint_provisioner.go:133]: @cjwmorgan-sol, this is probably no longer relevant since we are removing the GetEndpointProperties() entirely

File: internal/impl/constants/default_properties.go
Comment (oodigie) [reply internal/impl/constants/default_properties.go:76]: done

File: internal/impl/provisioner/endpoint_provisioner_impl.go
Comment (oodigie) [reply internal/impl/provisioner/endpoint_provisioner_impl.go:278]: Added the comment

File: internal/impl/core/endpoint_provisioner.go
Comment (oodigie) [reply internal/impl/core/endpoint_provisioner.go:174]: done

File: internal/impl/provisioner/endpoint_provisioner_impl.go
Comment (oodigie) [reply internal/impl/provisioner/endpoint_provisioner_impl.go:204]: Done

File: internal/impl/provisioner/endpoint_provisioner_impl.go
Comment (oodigie) [reply internal/impl/provisioner/endpoint_provisioner_impl.go:212]: Done

File: internal/impl/provisioner/endpoint_provisioner_impl.go
Comment (oodigie) [reply internal/impl/provisioner/endpoint_provisioner_impl.go:226]: Done

File: internal/impl/provisioner/endpoint_provisioner_impl.go
Comment (oodigie) [reply internal/impl/provisioner/endpoint_provisioner_impl.go:309]: Done

File: internal/impl/provisioner/endpoint_provisioner_impl.go
Comment (oodigie) [reply internal/impl/provisioner/endpoint_provisioner_impl.go:314]: Done

File: internal/impl/provisioner/endpoint_provisioner_impl.go
Comment (oodigie) [reply internal/impl/provisioner/endpoint_provisioner_impl.go:327]: Done

File: pkg/solace/config/endpoint_properties_test.go
Comment (oodigie) [reply pkg/solace/config/endpoint_properties_test.go:27]: You are right, it is actually a serialization/deserialization test. I checked again and the initial comment I made was for another file. I thought it was for something else. I have updated the float64() to uint64()

File: internal/impl/provisioner/endpoint_provisioner_impl_test.go
Comment (oodigie) [reply internal/impl/provisioner/endpoint_provisioner_impl_test.go:45]: Thanks, updated it now.

File: pkg/solace/endpoint_provisioner.go
Comment (oodigie) [reply pkg/solace/endpoint_provisioner.go:120]: Thanks, I have updated the comment @sdewilde

File: pkg/solace/config/endpoint_properties_test.go
Comment (sdewilde) [reply pkg/solace/config/endpoint_properties_test.go:27]: So then should we add all the other props here as well to ensure they can be serialized?

File: test/endpoint_provisioner_test.go
Comment (sdewilde) [inline test/endpoint_provisioner_test.go:94]: Does the user know that this is happening if they set it? In other APIs (JS comes to mind from recent dev) the call to provision is rejected if durability is set to false

File: test/endpoint_provisioner_test.go
Comment (sdewilde) [inline test/endpoint_provisioner_test.go:387]: Max length is not one of the options in the `validQueueTestCases` map above

File: test/endpoint_provisioner_test.go
Comment (sdewilde) [inline test/endpoint_provisioner_test.go:591]: `-1` and `256` are already tested at the top of this file, is this testing a different code path by not using the builder methods?

File: test/endpoint_provisioner_test.go
Comment (sdewilde) [inline test/endpoint_provisioner_test.go:995]: How is this test different from the previous? The part you are testing is when there is no queue so why is it different if there used to be one or not?

**Review by sdewilde — CHANGES_REQUESTED** [review]
Some test changes/questions

File: test/endpoint_provisioner_test.go
Comment (sdewilde) [reply test/endpoint_provisioner_test.go:94]: @oodigie this for sure needs to change, from the arch "Provisioning temporary endpoints through this new interface (without queue template trickery) will initially be disabled..."

File: pkg/solace/config/endpoint_properties_test.go
Comment (oodigie) [reply pkg/solace/config/endpoint_properties_test.go:27]: Okay sure, make sense. I will update this with the full properties list

File: test/endpoint_provisioner_test.go
Comment (cjwmorgan-sol) [reply test/endpoint_provisioner_test.go:103]: See the BeforeEach function line 38

Comment (sdewilde) [reply test/endpoint_provisioner_test.go:103]: That just calls build not start?

**Comment by cjwmorgan-sol** [comment]
Few things to note integration tests now take 8 more seconds compared to dev.
596.401 seconds current branch.
588.128 seconds last dev branch.
And Coverage went from 89.5% down to 89.1%.
I believe this branch it not up to date with the base branch [SOL-106478](https://github.com/SolaceDev/pubsubplus-go-client/tree/SOL-106478).
Please synchronize the branch.
Other then that the quick look I took is in line with @sdewilde comments. I'll take a closer look once the branch is up to date.

File: pkg/solace/config/endpoint_properties_test.go
Comment (oodigie) [reply pkg/solace/config/endpoint_properties_test.go:27]: Done

File: test/endpoint_provisioner_test.go
Comment (oodigie) [reply test/endpoint_provisioner_test.go:94]: From the ARCH (and ARCH review meetings) we decided to expose the `WithDurability()` method and its corresponding property in the properties map but default it to True. I am not sure if this has changed.

File: test/endpoint_provisioner_test.go
Comment (oodigie) [reply test/endpoint_provisioner_test.go:103]: @sdewilde, that's the whole point, it is unstarted/unconnected because the `Connect()` method was not called on it.

File: test/endpoint_provisioner_test.go
Comment (oodigie) [reply test/endpoint_provisioner_test.go:387]: I was following the Test plan document by QA - https://sol-jira.atlassian.net/wiki/spaces/CPD/pages/4031283346/API+s+Support+For+Removing+Durable+Endpoints+Test+Plan

File: test/endpoint_provisioner_test.go
Comment (sdewilde) [reply test/endpoint_provisioner_test.go:103]: So all the previous tests are also on a unstarted/unconnected service as well?

File: test/endpoint_provisioner_test.go
Comment (oodigie) [reply test/endpoint_provisioner_test.go:995]: I followed the Test plan document by QA - https://sol-jira.atlassian.net/wiki/spaces/CPD/pages/4031283346/API+s+Support+For+Removing+Durable+Endpoints+Test+Plan . And that is why I added a this comment **"// TestPlan TestCase Deprovision#8"** on the affected test case(s)

File: test/endpoint_provisioner_test.go
Comment (sdewilde) [reply test/endpoint_provisioner_test.go:387]: The test name says "queueName of max length" so that's why I was expecting to see a max length queue name

File: test/endpoint_provisioner_test.go
Comment (oodigie) [reply test/endpoint_provisioner_test.go:94]: Not sure that this is what we discussed in our ARCH review. We agreed to set it to True even when the false is passed to the method. We should agree which decision we want to take in this scenario.

File: test/endpoint_provisioner_test.go
Comment (sdewilde) [reply test/endpoint_provisioner_test.go:591]: But is it different than the previous tests with the same values, no sense in testing it twice if we don't need to

File: test/endpoint_provisioner_test.go
Comment (oodigie) [reply test/endpoint_provisioner_test.go:103]: yes, the ones outside the context() block.

File: test/endpoint_provisioner_test.go
Comment (sdewilde) [reply test/endpoint_provisioner_test.go:995]: The test plan says "Try to deprovision a queue(topic endpoint) which does not exists" why is it different if the queue existed at some point before or was de-provisioned by the API already?

**Comment by oodigie** [comment]
@cjwmorgan-sol, these are noted. I think the idea is for this branch to be merged into the feature base branch (SOL-106478) and not the other way around. We can merge this into the base branch and continue our review there before we merge that new PR into dev.

File: test/endpoint_provisioner_test.go
Comment (sdewilde) [reply test/endpoint_provisioner_test.go:94]: The test plan specifically calls our trying this with false and expecting an exception/error.  In JS we raised https://sol-jira.atlassian.net/browse/SOL-116163 because the error was incorrect but we still expect something

File: test/endpoint_provisioner_test.go
Comment (sdewilde) [reply test/endpoint_provisioner_test.go:103]: Sounds good, thanks for the explanation!

File: test/endpoint_provisioner_test.go
Comment (oodigie) [reply test/endpoint_provisioner_test.go]: I have updated the name @sdewilde

File: test/endpoint_provisioner_test.go
Comment (oodigie) [reply test/endpoint_provisioner_test.go:94]: @sdewilde, I will update the Go API to match returning an error like in SolclientJS. I will have to update the ARCH to state this.
cc: @cjwmorgan-sol

Comment (oodigie) [reply test/endpoint_provisioner_test.go]: Done

File: internal/impl/constants/default_properties.go
Comment (sdewilde) [reply internal/impl/constants/default_properties.go]: @cjwmorgan-sol I'll give you resolve power here, this looks good to me

File: internal/impl/provisioner/endpoint_provisioner_impl.go
Comment (cjwmorgan-sol) [inline internal/impl/provisioner/endpoint_provisioner_impl.go:73]: The golang api must not be ensuring the durability value as true that should be delegated to ccsmp. We can determine its boolean. But honestly I'm not sure this shuld be a config Property to begin with. It should be set for ccsmp and always true but that is defined by the default properties.

File: internal/impl/constants/default_properties.go
Comment (cjwmorgan-sol) [reply internal/impl/constants/default_properties.go]: I think we got a problem with the use of the public config.EndpointPropertyDurable, this property if not conifgurable should not be public.

**Review by cjwmorgan-sol — CHANGES_REQUESTED** [review]
There is public interface issues around config.EndpointPropertyDurable.
We'll need a resolution for this.

File: internal/impl/validation/validator.go
Comment (cjwmorgan-sol) [inline internal/impl/validation/validator.go:149]: As per the internal discussion. We'll keep the endpoint durability property with setter. But then we must defer to ccsmp for the validity of the property value. This only test golang needs to do is for whether the value is boolean or not. Why was this added?

File: pkg/solace/config/endpoint_properties.go
Comment (cjwmorgan-sol) [reply pkg/solace/config/endpoint_properties.go:66]: As discussed this is fine with the setter.

File: test/endpoint_provisioner_test.go
Comment (cjwmorgan-sol) [inline test/endpoint_provisioner_test.go:364]: This is only on cleanup right?
Do you need to assert the error? again? This is just cleanup code right?
Shouldn't the expectation be a nil error, or if this pure cleanup code why expect anything?

**Review by cjwmorgan-sol — CHANGES_REQUESTED** [review]
Please fix the failing test.
Other that this is looking good.

File: test/endpoint_provisioner_test.go
Comment (oodigie) [reply test/endpoint_provisioner_test.go]: fixed the unstable tests, removed the unnecessary assertion in the defer function.

**Review by sdewilde — APPROVED** [review]

**Review by cjwmorgan-sol — APPROVED** [review]
Looks good to me.
This is ready for merge to the feature branch.

**Pull request merged**
